include/nektar_interface/projection/algorithm_types.hpp

```diff
@@ -0,0 +1,269 @@
+#pragma once
+
+#include "constants.hpp"
+#include "device_data.hpp"
+#include "shapes.hpp"
+#include <sycl/sycl.hpp>
+#include <limits>
+#include <optional>
+
+#define ROUND_UP_TO_MULTIPLE(MULTIPLE, X)                                      \
+  (MULTIPLE) * ((((X) - 1) / (MULTIPLE)) + 1)
+#define ROUND_DOWN_TO_MULTIPLE(MULTIPLE, X) (MULTIPLE) * ((X) / (MULTIPLE))
+
+namespace NESO::Project {
+namespace Private {
+template <int nmode, typename Shape>
+static inline uint64_t total_local_slots() {
+  if constexpr (Shape::dim == 2) {
+    return Shape::template local_mem_size<nmode, 0>(1) +
+           Shape::template local_mem_size<nmode, 1>(1);
+  } else if constexpr (Shape::dim == 3) {
+    return Shape::template local_mem_size<nmode, 0>(1) +
+           Shape::template local_mem_size<nmode, 1>(1) +
+           Shape::template local_mem_size<nmode, 2>(1);
+  } else {
+    static_assert(false, "Only support 2 and 3 dimensional cell types");
+    // unreachable
+    return -1;
+  }
+}
+
+template <int nmode, typename T, typename Shape>
+static inline uint64_t calc_max_local_size(sycl::queue &queue,
+                                           int preferred_block_size) {
+  auto dev = queue.get_device();
+  // We subtract 1024 because cuda reserves that much per SM for itself
+  // Don't know if there is some "sycl" way to get this info
+  // TODO: Find the corresponding number for AMD and Intel
+  uint64_t local_mem_max =
```

Comment:
Personally I would assume uint64_t is std::size_t and just use std::size_t everywhere.